Add coding phase 2 blogpost for checks api plugin (#3577)
jenkins-infra/jenkins.io: content/blog/2020/07/2020-07-30-github-checks-api-plugin-coding-phase-2.adoc

---
layout: post
title: "GitHub Checks API Plugin Project - Coding Phase 2"
tags:
- github
- api
- plugins
- developer
- gsoc
- gsoc2020
author: XiongKezhi
description: "Google Summer of Code 2020 - GitHub Checks API Plugin project update for Coding Phase 2"

> opengraph:

Reviewer: Would be better to have a custom image. I can help with a draft

Author: that would be nice, thanks.

Reviewer: https://docs.google.com/presentation/d/1Q1PgNnRTgzBpVRXPqQo3PudzCa2eoc6_1_NRjFRMLrU/edit#slide=id.g8f0f0276ed_0_0 maybe?

  image: /images/gsoc/opengraph.png
---

Another great coding phase for link:https://www.jenkins.io/projects/gsoc/2020/projects/github-checks/[GitHub Checks API Project] ends!

> In this phase, we focused on consuming the checks API in two widely used plugin:

Reviewer 2: ```suggestion
In this phase, we focused on consuming the checks API in two widely used plugins:
```

* link:https://plugins.jenkins.io/warnings-ng/[Warnings NG Plugin] 
* link:https://plugins.jenkins.io/code-coverage-api/[Code Coverage API Plugin]

> Besides the external usage, we have also split the general checks API from its GitHub implementation and released both of them:

Reviewer 2: ```suggestion
Besides the external usage, we have also split the general checks API from its GitHub implementation and released both of the plugins:
```

== Warning Checks

The newly released https://github.com/jenkinsci/warnings-ng-plugin/releases/tag/warnings-ng-8.4.0[Warnings NG plugin 8.4.0] will use checks API to publish different check runs for different static analysis tools.
Without leaving GitHub, users are now able to see the analysis report they interested in.

image::/images/post-images/2020/07-github-checks-api-plugin-coding-phase-2/warning-checks.png[Warning Checks Summary]

On GitHub's conversation tab for each PR, users will see summaries for those checks like the screenshot above. The summaries will include:

* The status that indicates the quality gate
* The name of the analysis tool used
* A short message that indicates statistics of new and total issues

More fine-grained statistics can be found in the `Details` page.

image::/images/post-images/2020/07-github-checks-api-plugin-coding-phase-2/severity-statistics.png[Severity Statis]

Another practical feature is the annotation for specific lines of code. Users can now review the code alone with the annotations.

image::/images/post-images/2020/07-github-checks-api-plugin-coding-phase-2/annotations.png[Warning Annotations]

=== Try It

> In https://github.com/jenkinsci/warnings-ng-plugin/releases/tag/warnings-ng-8.4.0[Wanings NG plugin 8.4.0], the warning checks is set as a default feature. 

Reviewer 3: Can you make clear that it is the default for GitHub only? All other jobs will get a ˋNullPublisherˋ automatically. See comments in the warnings Gitter channel. So nobody needs to change anything in the jobs...

Therefore, you can get those checks for your own GitHub project just in a few steps:

1. Update Warnings NG plugin to 8.4.0

> 2. Install link:https://plugins.jenkins.io/checks-api/[Checks API plugin] and link:https://plugins.jenkins.io/github-checks/[GitHub Checks plugin] on your Jenkins instance

Reviewer 3: The checks api plugin is a dependency of the warnings plugin. So only the GitHub-Checks Plugin needs to be selected manually.

3. Use warnings-ng plugin in your Jenkinsfile for the multi-branch project or GitHub organization project, e.g.

[source, groovy]
----
node {
    stage ('Checkout') {
        checkout scm
    }

    stage ('Build and Static Analysis') {

>         withMaven {
>             sh 'mvn -V -e clean verify -Dmaven.test.failure.ignore'
>         }

Reviewer 2: should be unneeded
```suggestion
        sh 'mvn -V -e clean verify -Dmaven.test.failure.ignore'
```

        recordIssues tools: [java(), javaDoc()], aggregatingResults: 'true', id: 'java', name: 'Java'
        recordIssues tool: errorProne(), healthy: 1, unhealthy: 20
        recordIssues tools: [checkStyle(pattern: 'target/checkstyle-result.xml'),
            spotBugs(pattern: 'target/spotbugsXml.xml'),
            pmdParser(pattern: 'target/pmd.xml'),
            cpd(pattern: 'target/cpd.xml')],
            qualityGates: [[threshold: 1, type: 'TOTAL', unstable: true]]
    }
}
----

_For more about the pipeline usage of warnings-ng plugin, please see link:https://github.com/jenkinsci/warnings-ng-plugin/blob/master/doc/Documentation.md#pipeline-configuration[the official documentation]._

> However, if you don't want to publish the warnings to GitHub, you can definitely disable it by using a boolean field `skipPublishingChecks`, e.g.

Reviewer 2: ```suggestion
However, if you don't want to publish the warnings to GitHub, you can disable it by adding `skipPublishingChecks: true`.
```

[source, groovy]
----
recordIssues enabledForFailure: true, tools: [java(), javaDoc()], skipPublishingChecks: true
----

== Coverage Checks

The coverage checks are achieved by consuming the API in link:https://plugins.jenkins.io/code-coverage-api/[Code Coverage API plugin].

> First, in the conversation tab of a PR, users will be able to see the summary about the coverage and the comparing with previous builds.

Reviewer 2: ```suggestion
First, in the conversation tab of a PR, users will be able to see the summary about the coverage and the coverage difference compared to previous builds.
```

image::/images/post-images/2020/07-github-checks-api-plugin-coding-phase-2/coverage-summary.png[Coverage Summary]

> Besides, the `Details` page will contain some other things:

Reviewer 2: ```suggestion
The `Details` page will contain some other things:
```

> * The links to the reference build, including the target branch build from the master branch and the last successful build from this branch

Reviewer 2: ```suggestion
* Links to the reference build, including the target branch build from the master branch and the last successful build from this branch
```

* Coverage healthy score (the default value is 100% if the threshold is not configured)
* Coverages and trends of different types in table format

image::/images/post-images/2020/07-github-checks-api-plugin-coding-phase-2/coverage-details.png[Coverage Details]

The link:https://github.com/jenkinsci/code-coverage-api-plugin/pull/169[pull request] for this feature will soon be merged and will be included in the next release of Coverage Checks API plugin. After that, you can use it by adding the below section to your pipeline script:

[source, groovy]
----
node {
    stage ('Checkout') {
        checkout scm
    }

    stage ('Line and Branch Coverage') {

>         withMaven {
>             sh "mvn -V -U -e jacoco:prepare-agent test jacoco:report -Dmaven.test.failure.ignore"
>         }

Reviewer 2: the jacoco step isn't actually needed if you follow the docs in the plugin for configuring maven, https://plugins.jenkins.io/code-coverage-api/ (search jacoco), not sure if you want to rework this a bit for that?
```suggestion
        sh "mvn -V -U -e jacoco:prepare-agent test jacoco:report -Dmaven.test.failure.ignore"
```

        publishCoverage adapters: [jacocoAdapter('**/*/jacoco.xml')], sourceFileResolver: sourceFiles('STORE_ALL_BUILD')
    }
}
----

Like the warning checks, you can also disable the coverage checks by setting the field `skipPublishingChecks`, e.g.

[source, groovy]
----
publishCoverage adapters: [jacocoAdapter('**/*/jacoco.xml')], sourceFileResolver: sourceFiles('STORE_ALL_BUILD'), skipPublishingChecks: true
----

== Next Phase

In the next phase, we will turn our attention back to link:https://plugins.jenkins.io/checks-api/[Checks API Plugin] and link:https://plugins.jenkins.io/github-checks/[GitHub Checks Plugin]:

> * Pipeline support will be added, so users can publish checks directly in a pipeline script without requiring a consumer plugin that supports the checks.

Reviewer: "will be added in future versions" ?

* Re-run request for Jenkins build through Checks API will be added as well. 

> Hopefully, we will also release the 1.0.0 version for both plugins.

Reviewer: Reference phase 3? I'd guess hope only applies to the timeline, not to a fact of the release

> Lastly, it is exciting to inform that we are currently making the checks feature available on link:https://ci.jenkins.io[ci.jenkins.io] for hosted plugins, please see link:https://issues.jenkins-ci.org/browse/INFRA-2694[INFRA-2694] for more.

Reviewer 2: ```suggestion
Lastly, it is exciting to inform that we are currently making the checks feature available on link:https://ci.jenkins.io[ci.jenkins.io] for all plugins hosted in the jenkinsci GitHub organisation, please see link:https://issues.jenkins-ci.org/browse/INFRA-2694[INFRA-2694] for more details.
```

> Lastly, it is exciting to inform that we are currently making the checks feature available on link:ci.jenkins.io[ci.jenkins.io] for hosted plugins, please see link:https://issues.jenkins-ci.org/browse/INFRA-2694[INFRA-2694] for more.

Reviewer: ci.jenkibs.io link has no https prefix

---
layout: post
title: "GitHub Checks API Plugin Project - Coding Phase 2"
tags:
- github
- api
- plugins
- developer
- gsoc

> - gsoc2020

Reviewer: Would be nice to add a description and an opengraph image

On GitHub's conversation tab for each PR, users will see summaries for those checks like the screenshot above. The summaries will include:

* The status that indicates the quality gate
* The name of the analysis tool used
* A short message that indicates statistics of new and total issues

More fine-grained statistics can be found in the `Details` page.

image::/images/post-images/2020/07-github-checks-api-plugin-coding-phase-2/severity-statistics.png[Severity Statis]

Another practical feature is the annotation for specific lines of code. Users can now review the code alone with the annotations.

image::/images/post-images/2020/07-github-checks-api-plugin-coding-phase-2/annotations.png[Warning Annotations]

=== Try It

In https://github.com/jenkinsci/warnings-ng-plugin/releases/tag/warnings-ng-8.4.0[Wanings NG plugin 8.4.0], the warning checks is set as a default feature. 
Therefore, you can get those checks for your own GitHub project just in a few steps:

1. Update Warnings NG plugin to 8.4.0
2. Install link:https://plugins.jenkins.io/checks-api/[Checks API plugin] and link:https://plugins.jenkins.io/github-checks/[GitHub Checks plugin] on your Jenkins instance

> 3. Configure warnings-ng plugin as usual for a multi-branch project or a GitHub organization project

Reviewer 2: I would provide an example rather than saying as usual

=== Pipeline Usage:

> Pipeline users can control this feature through a boolean field `skipPublishingChecks`, e.g.

Reviewer 2: I would provide a regular example before showing how to disable it

```
recordIssues tool: errorProne(), healthy: 1, unhealthy: 20, skipPublishingChecks: false
```

== Coverage Checks

The coverage checks are achieved by consuming the API in link:https://plugins.jenkins.io/code-coverage-api/[Code Coverage API plugin].
First, in the conversation tab of a PR, users will be able to see the summary about the coverage and the comparing with previous builds.

image::/images/post-images/2020/07-github-checks-api-plugin-coding-phase-2/coverage-summary.png[Coverage Summary]

Besides, the `Details` page will contain some other things:

* The links to the reference build, including the target branch build from the master branch and the last successful build from this branch
* Coverage healthy score (the default value is 100% if the threshold is not configured)
* Coverages and trends of different types in table format

image::/images/post-images/2020/07-github-checks-api-plugin-coding-phase-2/coverage-details.png[Coverage Details]

> The link:https://github.com/jenkinsci/code-coverage-api-plugin/pull/169[pull request] for this feature will soon be merged and will be included in the next release of Coverage Checks API plugin.

Reviewer 2: Should still include how to use this, I.e an example pipeline script

> == Next Phase

Reviewer: Would it make sense to mention that we are also working on making the feature available on ci.jenkins.io for plugin maintainers? https://issues.jenkins-ci.org/browse/INFRA-2694

Once completed, it would be a major feature to showcase IMHO. CC @timja @uhafner

In the next phase, we will turn our attention back to link:https://plugins.jenkins.io/checks-api/[Checks API Plugin] and link:https://plugins.jenkins.io/github-checks/[GitHub Checks Plugin]. 
Pipeline support will be added: users can publish checks directly in a pipeline script without requiring a consumer plugin that supports the checks.
Re-run request for Jenkins build through Checks API will be added as well. 

> Hopefully, we will also release the first official version for both plugins.

Reviewer 2: We have already released official versions, 1.0 release?
